src/nfa/nfa.cc

```diff
@@ -312,6 +312,136 @@ struct StackLevel {
     };
 };
 
+struct NodeData {
+    Post::const_iterator post_it;
+    Post::const_iterator post_end;
+    StateSet::const_iterator targets_it{};
+    StateSet::const_iterator targets_end{};
+    unsigned long index;
+    unsigned long lowlink;
+    bool initilized;
+    bool on_stack;
+
+    NodeData() = default;
+
+    NodeData(State q, const Delta & delta, unsigned long index) 
+        : post_it(delta[q].cbegin()), post_end(delta[q].cend()), index(index), lowlink(index), initilized(true), on_stack(true) {
+        if (post_it != post_end) {
+            targets_it = post_it->cbegin();
+            targets_end = post_it->cend();
+        }
+    };
+
+    // TODO: this sucks. In fact, if you want to check that you have the last sucessor, you need to 
+    // first align the iterators.
+    // TODO: this is super-ugly. If we introduce Post::transitions iterator, this could be much easier.
+    void align_succ() {
+        while (this->post_it != this->post_end && this->targets_it == this->targets_end) {
+            if (this->targets_it == this->targets_end) {
+                ++this->post_it;
+                if (this->post_it != this->post_end) {
+                    this->targets_it = this->post_it->cbegin();
+                    this->targets_end = this->post_it->cend();
+                }
+            } 
+        }
+    }
+
+    State get_act_succ() {
+        align_succ();
+        return *(this->targets_it);
+    }
+
+    void move_next_succ() {
+        if(this->post_it == this->post_end) {
+            return;
+        }
+        this->targets_it++;
+    }
+
+    bool is_end() {
+        align_succ();
+        return this->post_it == this->post_end;
+    }
+};
+
+BoolVector Nfa::get_useful_states_tarjan() const {
+    BoolVector reached_and_reaching(this->size(),false); 
+    std::vector<NodeData> node_info(this->size());
```

Comment:
`NodeData` contains `bool` and `unsigned long` variables which are not initialized, I think this will lead to UB.

Reply:
I don't get it

Reply:
You are creating vector of `NodeData`, and they use default constructor. Will that initialize the `bool` members to `false`? You assume it does, on line 383 you expect `false`.

Reply:
I expected that it is initialized to `false` by default. Is it the case? @Adda0

Reply:
The magic around default initialization, value initialization, explicit initialization, and more is complicated. For a brief overview, see example in [LearnCpp article on default initialization](https://www.learncpp.com/cpp-tutorial/default-member-initialization/). The best approach is to initialize everything as explicitly as possible, at least with `{}` in member variable declaration. In this case, both I and Clang-tidy believe that there is a possibility of UB with the default constructor in the following members: `index`, `lowlink`, `initilized`, `on_stack`.

Another great explanation is in the following [blog post](https://arne-mertz.de/2015/08/new-c-features-default-initializers-for-member-variables/). Initialize everything everywhere and you will be safe. If your member variable is not a class, it will not get initialized if you use only `NodeData node_data;`.

Reply:
Great, thanks!